### PR TITLE
feat: support add plugins for specified environment

### PR DESCRIPTION
**Pull request by 9aoy**
## Summary

Support add plugins for specified environment.

For example, enable the react plugin only in the web environment:

```ts title=rsbuild.config.ts
import { pluginReact } from '@rsbuild/plugin-react';

export default {
  environments: {
    web: {
      output: {
        target: 'web',
      },
      plugins: [pluginReact()],
    },
    node: {
      output: {
        target: 'node',
      },
    },
  },
};
```

TODO: Some plugins do not yet support single-environment registration because they use `api.modifyRsbuildConfig` hook to modify the global configuration.

## Related Links

https://github.com/web-infra-dev/rsbuild/issues/2620

## Checklist

- [ ] Tests updated (or not required).
- [ ] Documentation updated (or not required).

### CODE REVIEW

File: packages/core/src/initHooks.ts [hunk 1]
```diff
@@ -22,6 +25,75 @@ import type {
...
+          environmentMeta,
+          handler: cb.handler,
+        });
+      }
+    };
+
+  const call =
+    (environment?: string) =>
+    async (...args: Parameters<Callback>) => {
+      const params = args.slice(0) as Parameters<Callback>;
```
Comment (chenjiahan) [inline packages/core/src/initHooks.ts:68]: Do we need `slice(0)`?

File: packages/core/src/initHooks.ts [hunk 2]
```diff
@@ -22,6 +25,75 @@ import type {
...
+      } else {
+        defaultGroup.push({
+          environmentMeta,
+          handler: cb.handler,
+        });
+      }
+    };
+
+  const call =
+    (environment?: string) =>
```
Comment (chenjiahan) [inline packages/core/src/initHooks.ts:66]: Why does `tap` accept `environmentMeta?: EnvironmentMeta`, but `call` accepts `environment?: string`, can we make them consistent.

File: packages/core/src/types/plugin.ts [hunk 3]
```diff
@@ -90,16 +103,39 @@ export type ModifyWebpackConfigFn = (
...
+  getPlugins: (options?: {
+    /** Get the plugins in the specified environment. If environment is not specified, get the global plugins.  */
+    environment: string;
+  }) => RsbuildPlugin[];
   addPlugins: (
     plugins: Array<RsbuildPlugin | Falsy>,
-    options?: { before?: string },
+    options?: {
+      before?: string;
+      /** Add a plugin for the specified environment. If environment is not specified, it will be registered as a global plugin (effective in all environments)  */
```
Comment (chenjiahan) [inline packages/core/src/types/plugin.ts:119]: We should avoid having too many characters on a line

File: website/docs/zh/guide/advanced/environments.mdx [hunk 4]
```diff
@@ -100,6 +100,32 @@ export default {
 };
 ```
 
+## 为指定环境添加插件
+
+通过 [plugins](/config/plugins) 字段配置的插件支持在所有环境下运行，如果你希望某个插件仅在指定环境下运行时，将该插件配置在指定 environment 下即可。
+
+如，仅在 web 环境下开启 react 插件：
```
Comment (chenjiahan) [inline website/docs/zh/guide/advanced/environments.mdx:107]:
```suggestion
例如，仅在 web 环境下开启 React 插件：
```

File: website/docs/zh/guide/advanced/environments.mdx [hunk 5]
```diff
@@ -100,6 +100,32 @@ export default {
 };
 ```
 
+## 为指定环境添加插件
+
+通过 [plugins](/config/plugins) 字段配置的插件支持在所有环境下运行，如果你希望某个插件仅在指定环境下运行时，将该插件配置在指定 environment 下即可。
```
Comment (chenjiahan) [inline website/docs/zh/guide/advanced/environments.mdx:105]:
```suggestion
通过 [plugins](/config/plugins) 字段配置的插件支持在所有环境下运行，如果你希望某个插件仅在指定环境下运行时，将该插件配置在特定 `environment` 下即可。
```

File: website/docs/en/guide/advanced/environments.mdx [hunk 6]
```diff
@@ -100,6 +100,32 @@ export default {
 };
 ```
 
+## Add plugins for specified environment
+
+Plugins configured through the [plugins](/config/plugins) field support running in all environments. If you want a plugin to run only in a specified environment, you can configure the plugin in the specified environment.
```
Comment (chenjiahan) [inline website/docs/en/guide/advanced/environments.mdx:105]:
```suggestion
Plugins configured through the [plugins](/config/plugins) field support running in all environments. If you want a plugin to run only in a specified environment, you can configure the plugin in the specified `environment`.
```

File: website/docs/en/guide/advanced/environments.mdx [hunk 7]
```diff
@@ -100,6 +100,32 @@ export default {
...
+
+For example, enable the react plugin only in the web environment:
```
Comment (chenjiahan) [inline website/docs/en/guide/advanced/environments.mdx:107]:
```suggestion
For example, enable the React plugin only in the web environment:
```

File: packages/core/src/initHooks.ts (hunk 1 above)
Comment (9aoy) [reply packages/core/src/initHooks.ts]: removed, just copy from `createAsyncHook`...

**Review by chenjiahan — APPROVED** [review]
Awesome 🎉

**Pull request merged**